libs/datasets/timeseries.py

```diff
@@ -1031,10 +1031,84 @@ def drop_regions_without_population(
 LOCATION_ID_AGG = "location_id_agg"
 
 
+@dataclass(frozen=True)
+class AverageStaticWeightAggregation:
+    """Information about a timeseries field which is aggregated using a weighted average."""
+
+    # timeseries field that gets aggregated using a weighted average
+    field: FieldName
+    # static field that used to produce the weights
+    scale_factor: FieldName
+
+
+AGGREGATIONS = (
+    AverageStaticWeightAggregation(CommonFields.TEST_POSITIVITY, CommonFields.POPULATION),
+    AverageStaticWeightAggregation(
+        CommonFields.ALL_BED_TYPICAL_OCCUPANCY_RATE, CommonFields.MAX_BED_COUNT
+    ),
+    AverageStaticWeightAggregation(CommonFields.ICU_TYPICAL_OCCUPANCY_RATE, CommonFields.ICU_BEDS),
+)
+
+
+def _apply_scaling_factor(
+    df_in: pd.DataFrame,
+    scale_factors: pd.DataFrame,
+    aggregations: Sequence[AverageStaticWeightAggregation],
+) -> pd.DataFrame:
+    """Returns a copy of timeseries with some fields scaled according to `aggregations`.
+
+    Args:
+        df_in: Input un-aggregated timeseries or static data
+        static_in: Input un-aggregated static data, used to find scaling factor
+        static_agg: Aggregated static data
+        location_id_map: Maps from un-aggregated location_id to aggregated location_id
+        aggregations: Describes the fields to be scaled
+        """
+    assert df_in.index.names in (
+        [CommonFields.LOCATION_ID, CommonFields.DATE],
+        [CommonFields.LOCATION_ID],
+    )
+
+    # Scaled fields are modified in-place
+    df_out = df_in.copy()
+
+    for agg in aggregations:
+        if agg.field in df_in.columns and agg.scale_factor in scale_factors.columns:
+            df_out[agg.field] = df_out[agg.field] * scale_factors[agg.scale_factor]
+
+    return df_out
+
+
+def _find_scale_factors(
+    aggregations, location_id_map, static_agg, static_in, location_ids
```

Comment:
Could you add types here? would be helpful for context

Reply:
Done